### PR TITLE
Add relations cache busting

### PR DESCRIPTION
**Pull request by julesrenaud**
@patrixr Here's an implementation of an option to bust related caches on POST, PUT and DELETE calls (as per proposed in #13). It piggybacks on the existing `clearCache` helper (that way, if enabled, the related caches will always be busted) and is enabled via a `clearRelatedCache` middleware option.

What it does basically is:

1. Gather the related models by [inspecting the updated model's relations to other models](https://github.com/principalstudio/strapi-middleware-cache/blob/15e756d07d721861178e28a5070b2ba4a8f47667/lib/index.js#L101-L112).
2. It then [goes looking for components that have the updated model as a relation](https://github.com/principalstudio/strapi-middleware-cache/blob/15e756d07d721861178e28a5070b2ba4a8f47667/lib/index.js#L114-L150). We go two levels deep in the components to cover use cases such as a relation in a component that is then included in a dynamic zone.
3. It then [gathers the models where the related components are used](https://github.com/principalstudio/strapi-middleware-cache/blob/15e756d07d721861178e28a5070b2ba4a8f47667/lib/index.js#L152-L172).
4. Finally, [it constructs cache keys for the related models](https://github.com/principalstudio/strapi-middleware-cache/blob/15e756d07d721861178e28a5070b2ba4a8f47667/lib/index.js#L174-L183) and adds them to the existing `rexps` array to be busted alongside the updated model's cache.

It works well, I've tested it in a working Strapi installation. I'm not 100% happy with the structure, there's some repeated code. I've tried to spin off the functionality in a separate helper but it became too convoluted and hard to follow. At least the way it is now it's easy to follow albeit a little clumsy.

Let me know what you think! I can write tests if you agree with the approach, I'd have to add dummy models in the `strapi` instance in the tests, as it lacks the `strapi.models`, `strapi.plugins` and `strapi.components` objects needed to gather the related models.

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
> It then goes looking for components that have the updated model as a relation. We go two levels deep in the components to cover use cases such as a relation in a component that is then included in a dynamic zone.

Great choice :wink:

**Comment by julesrenaud** [comment]
> > It then goes looking for components that have the updated model as a relation. We go two levels deep in the components to cover use cases such as a relation in a component that is then included in a dynamic zone.
> 
> Great choice 😉

I rely heavily on relations embedded in components so I already had such use cases to test against. @derrickmehaffy it's actually you that gave me the insight yo use relations that way :)

**Comment by patrixr** [comment]
Hi @julesrenaud 

Thanks for this ! I think the approach would work nicely, but I think maybe we should move a bit of that code out to a helper, maybe something like:

```javascript
        if (clearRelatedCache) {
          const relatedModels = getRelatedModels(model); // move the population code to its own function

          // prep related models to have their whole cache busted
          _.each(relatedModels, (relatedModel, key) => {
              const relatedModelSlug = relatedModel.kind === 'collectionType'
                ? relatedModel.collectionName
                : key;

              rexps.push(new RegExp(`^/${relatedModelSlug}`))  // Bust all entries from models that have relations with that model
          })
        }
```

This would:
- Keep the cache busting / cache key code in the same file and function
- Extract the relational aspect of the code to its own file/helper, which can maintain the current step-by-step approach that you're using

Regarding tests, do add any mocks needed. I am aware it is quite a hacky setup currently, we're not (yet) striving for perfection but adding some basic tests to smoke out regression will go a long way :)

Cheers 🍺  !

**Comment by julesrenaud** [comment]
@patrixr agreed, I'll go ahead and implement these changes

**Comment by julesrenaud** [comment]
@patrixr @derrickmehaffy the populating code has been moved to a separate helper as suggested. I've retained for...in and for...of loops in the helper to leverage execution termination. The `strapi` object in the tests has also been mocked to match a real-use data structure. I've added the `researcher` collection type which is directly related to the `academy` model and is related to the `homepage` model via components (2-level deep). The tests are successful on my end, the cache is busted for the related model in both cases.

**Comment by patrixr** [comment]
@julesrenaud 

Looks good to me ! Well done on the specs !

Few small things:
- Could you bump the version to `1.4.0` ? (for the autopublish when we merge)
- Could you edit the README to mention this feature ? I would do it myself but since my understanding of the logic is only rudimentary I think you'd do a better job at describing it

Thanks ! 🤜 🤛

**Comment by julesrenaud** [comment]
@patrixr there you go!

**Comment by patrixr** [comment]
Great ! Thank you (again) for the contribution :)

Version 1.4.0 is now on npm
